### PR TITLE
Always show end line of multiline annotations

### PR DESCRIPTION
**Pull request by estebank**
```rust
error[E0046]: not all trait items implemented, missing: `Item`
  --> $DIR/issue-23729.rs:20:9
   |
20 |           impl Iterator for Recurrence {
   |  _________^ starting here...
21 | |             //~^ ERROR E0046
22 | |             //~| NOTE missing `Item` in implementation
23 | |             //~| NOTE `Item` from trait: `type Item;`
...  |
36 | |             }
37 | |         }
   | |_________^ ...ending here: missing `Item` in implementation
   |
   = note: `Item` from trait: `type Item;`
```

instead of

```rust
error[E0046]: not all trait items implemented, missing: `Item`
  --> $DIR/issue-23729.rs:20:9
   |
20 |         impl Iterator for Recurrence {
   |         ^ missing `Item` in implementation
   |
   = note: `Item` from trait: `type Item;`
```

### CODE REVIEW

**Comment by nrc** [comment]
r? @jonathandturner 

My personal opinion - this looks like an improvement - I find the new error more visually attractive, however, I don't think it adds anything to the explanatory power of the error message and everything else being equal, I prefer the more compact form (disclaimer, I'm only looking at the example in the comment here, perhaps there are other errors where this is more advantageous?)

**Comment by estebank** [comment]
@nrc I'm of the opinion that errors like the example should look like this:

```rust
error[E0046]: not all trait items implemented, missing: `Item`
  --> $DIR/issue-23729.rs:20:9
   |
20 |           impl Iterator for Recurrence {
   |           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^ missing `Item` in implementation
   |
   = note: `Item` from trait: `type Item;`
```

(as proposed in PR #38328, and issues #35965 and #38246) and leave multiline spans only for situations where they are actually useful, like:


```rust
  --> <anon>:2:5
   |
2  |       fn foo(
   |  _____^ starting here...
3  | |         a: usize,
4  | |         b: usize,
5  | |         c: usize)
6  | |         -> usize {
   | |________________^ ...ending here: foo definition span
```

**Comment by nrc** [comment]
@estebank perhaps I misunderstood - from the first comment I thought this PR did exactly the opposite of that?

**Comment by estebank** [comment]
@nrc this PR makes affects every multiline span.

A separate aim I have is to reduce how many multiline spans are actually used in diagnostics. I believe that every `Item` element should have a `diagnostic_span` separate from the `span` that is used now, that only covers the beginning. This would leave multiline spans only to be used for signatures that span multiple lines and blocks, which are used also for lifetimes, where the beginning and end is relevant. That would make multiline spans less frequent than they are now. Until that change is done though, this change would indeed make things noisier.

**Comment by estebank** [comment]
An example of a valid multiline span usage IMO:

```rust
note: the anonymous lifetime #1 defined on the body at 30:83...
  --> file.rs:30:84
   |
30 |       pub fn child_at_index_mut(&mut self, index: &NodeChildIndex) -> &mut NodeChild {
   |  ____________________________________________________________________________________^ starting here...
31 | |         self.children[index.index].borrow_mut()
32 | |     }
   | |_____^ ...ending here
note: ...does not necessarily outlive the lifetime 'a as defined on the body at 30:83
  --> file.rs:30:84
   |
30 |       pub fn child_at_index_mut(&mut self, index: &NodeChildIndex) -> &mut NodeChild {
   |  ____________________________________________________________________________________^ starting here...
31 | |         self.children[index.index].borrow_mut()
32 | |     }
   | |_____^ ...ending here
```

**Comment by jonathandturner** [comment]
@estebank - for that example do you have what we show today?

**Comment by estebank** [comment]
@jonathandturner, on the description of the PR there's an example of after and before. If a span is under 8 lines, we show the entire span, otherwise we only point at the first character.

This PR changes this to show in full spans shorter or equal to 7 lines, and longer spans show the first 4 lines and last two. (I thought about showing the last line only, but that would in most cases be just `}`, which isn't that illuminating.)

**Comment by jonathandturner** [comment]
Sure, seems fine.  I agree with @nrc that it mostly is a visual fix, which is fine.

@bors r+

File: src/librustc_errors/emitter.rs [hunk 1]
```diff
@@ -483,16 +461,15 @@ impl EmitterWriter {
...
-        // on a different line as the underline.
+        // Write the vertical lines for labels that are on a different line as the underline.
         //
         // After this we will have:
         //
         // 2 |   fn foo() {
         //   |  __________
         //   | |    |
         //   | |
-        // 3 | |
```
Comment (bjorn3) [inline src/librustc_errors/emitter.rs]: Why is this `|` deleted?

**Pull request merged**